### PR TITLE
look up constructors by type name in depwarn. fixes #21972

### LINKED ISSUE #21972: depwarn in constructors keeps warning
The logic for checking the binding is wrong for constructors so
```julia
struct TypeA
    function TypeA()
        Base.depwarn("blabla", :TypeA)
        new()
    end
end

julia> TypeA()
WARNING: blabla
Stacktrace:
 [1] depwarn(::String, ::Symbol) at ./deprecated.jl:64
 [2] TypeA() at ./REPL[6]:3
 [3] eval(::Module, ::Any) at ./boot.jl:235
 [4] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
 [5] macro expansion at ./REPL.jl:97 [inlined]
 [6] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
while loading no file, in expression starting on line 0
TypeA()

julia> TypeA()
WARNING: blabla
Stacktrace:
 [1] depwarn(::String, ::Symbol) at ./deprecated.jl:64
 [2] TypeA() at ./REPL[6]:3
 [3] eval(::Module, ::Any) at ./boot.jl:235
 [4] eval_user_input(::Any, ::Base.REPL.REPLBackend) at ./REPL.jl:66
 [5] macro expansion at ./REPL.jl:97 [inlined]
 [6] (::Base.REPL.##1#2{Base.REPL.REPLBackend})() at ./event.jl:73
while loading no file, in expression starting on line 0
TypeA()
```

### PR DESCRIPTION
**Pull request by JeffBezanson**

### CODE REVIEW

**Comment by tkelman** [comment]
we have deprecation tests now, can this be tested?

**Comment by JeffBezanson** [comment]
Ah, the tests are being run with `--depwarn=error`, so I can't directly test giving a deprecation warning. I guess I should run it in a separate process?

**Comment by tkelman** [comment]
I thought that's what the depreciation tests were doing, but maybe not yet. Maybe have a deprecation_exec.jl so we can test running it with each value of the flag, no vs warn vs error

**Comment by JeffBezanson** [comment]
Added `deprecation_exec.jl`.

File: test/deprecation_exec.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+# issue #21972
+struct T21972
+    @noinline function T21972()
+        Base.depwarn("something", :T21972)
+        new()
+    end
+end
+
+@testset "@deprecate" begin
```
Comment (tkelman) [inline test/deprecation_exec.jl]: hm, both this testset and `in anonymous at <missing>` are printing here and escaping both the runtests.jl printing and `@test_warn` capturing output

**Pull request merged**